### ISSUE TITLE
[Discussion] Pipelines in merging new benchmark results.

### ISSUE THREAD

**Issue by XericZephyr** [deepchem/deepchem#433]
Hi, @rbharath , 

I want to discuss a process for merging new benchmark results here. For example, if I want to merge my result into your benchmark, what kind of process should I follow? Should I provide a gist/github repo to reproduce my results? Should I ask at least one more DeepChem Admin to reproduce my result for confirmation?

Thanks.

**Comment by rbharath** [comment]
@XericZephyr This is a great question. I'll write up an answer here, but we should formalize and incorporate into the Docs for DeepChem once agreed upon.

1. In order to be run against MoleculeNet, proposed models should be DeepChem compatible. In particular, it means the proposed model should be a class which inherits from `dc.model.Model`. This abstract class follows APIs similar to commonly used python machine-learning classes from Scikit-Learn, so the process shouldn't be too onerous. One important note is that the model object will need to consume `dc.data.Dataset` objects. Check out the `examples/` folder for many examples of how to process datasets into `dc.data.Dataset`s.
2. Add the newly proposed model into a new file in `deepchem/models`. This file should contain links to associated papers and a high level explanation of the algorithm in docstrings. Make sure to add some basic unit tests for this new class to `deepchem/models/tests`!
3. Add the proposed model to `deepchem/examples/benchmark.py`. This file is used to run all MoleculeNet benchmarks. Use `benchmark.py` to run your proposed model on MoleculeNet datasets. Make sure you're setting the random seed to a fixed value!
4. Add reported numbers on MoleculeNet datasets to `README.md`. Report the commands used to produce these numbers in PR.
5. Make a pull-request to DeepChem with all these changes. Label this pull-request with tag `[MoleculeNet]`. Make sure the pull-request passes all unit tests.
6. One of the DeepChem devs will check the PR for basic correctness. If everything looks good, the admin will use the commands reported in PR to independently run the new model on MoleculeNet. If results are reproducible (to +/- 1%), then PR will be accepted, and the new model will be integrated into MoleculeNet benchmarks/results.

@miaecle Does this process sound reasonable? Is there anything else we should add on here?

**Comment by XericZephyr** [comment]
Thanks, @rbharath. That sounds really good to me. 

My only concern is sometimes if the proposed model is too complicated, e.g., might need to bring in many new dependencies, or utilizing some private repo, how could we add it into the models? And I believe if the number of merging requests increases, it will become extremely hard to maintain those code.

My experience was, instead of directly including the method implementation, we let the user create their own repo/gist for their **method**.

It might be a better idea for DeepChem to provide a **environment** or playground for those authors. A famous example of this model is [OpenAI gym](https://github.com/openai/gym). It provides the users with a **data input API** and **evaluation API**. Then the users are free to create their own method based on the framework. If the evaluation looks good, they can even **upload their result to show off on the scoreboard**.

Just a bit of my personal thoughts. How do you guys think?

**Comment by rbharath** [comment]
@XericZephyr These are great comments. I'd suggest for now let's follow the protocol I outlined above, but you raise good points about complex models and private repos. We will need to do some proper API design to make it feasible for, say, commercial entities to submit models to MoleculeNet without having to open-source their code.

@peastman Would you have any thoughts on API design here? We need to design a general API which enables MoleculeNet submissions without necessarily having to merge the code directly into DeepChem's github repo.

**Comment by XericZephyr** [comment]
OK. I just took a look at the `models` folder, it looks super good.

But I will need some of you help with my recent work. It mostly looks like an autoencoder model but it is actually a mixture of sklearn, tensorflow and autoencoder. So I have no idea where to place it.

Our code is implemented in tensorflow, without keras. And I will surely download some pretrained network file and some *dictionary file*. I will need to do a model evaluation to convert each molecule to a fingerprint. Then some sklearn classifier/regressor will be used to produce the final prediction.

So for my case, how can I get started?  😂

**Comment by rbharath** [comment]
Ah, I see. This is a little more complicated than I thought :-)

For downloading, I think you can `wget` the weights during model construction. Some of the models in `dc.dock` do this already.

I'd recommend converting the molecules into fingerprints within your `fit()` method. After the fingerprints are all generated, you can build the sklearn model.

A good strategy might be to start a new PR for DeepChem and mark it with the tag `[WIP]`. I can review the code directly there and make concrete suggestions to help the integration process.

**Comment by XericZephyr** [comment]
Thanks, @rbharath  for the quick response. Sure I will absolutely create a PR for this. 

But as a first step, I would like to show you guys my pipeline in a small dataset from DeepChem. (I might show it in an external gist, in a few days, since I am really not familiar with the `model` framework, sorry about this.) You guys might have some better ideas what I can do next.

A bit more adding about the API design, @rbharath, @peastman.
After the gist, I think you might also have the idea about working out the API, I might use some of them.

**Comment by rbharath** [comment]
Got some useful comments on the gitter from @patrickhop. We will want to make it so that closed-source users can run MoleculeNet for themselves locally, but I think we'll need to restrict "official" MoleculeNet additions to open-source only. Since the labels for all test sets are available publicly, unscrupulous actors can cheat to build perfectly accurate MoleculeNet classifiers/regressors. With open-source code, DeepChem devs can do code-review to make sure no cheating is happening.

**Comment by mrwns** [comment]
just a few things to think about:

there might be a mechanism needed to stop an explosion of dependencies:

For example, I want to benchmark an xgboost model, and another neural net based on pytorch. Let's assume the results are inferior to the graph convolutions, so these models would never be used by any user in practive.

However, to run the unit tests, the deepchem user would have to install all those unneeded dependencies. It could be better to keep the benchmarks in a separate "molnet" repo, and if a contributed models performs extraordinarily good, it could be merged into deepchem.

Additionally, the deepchem repo is already very large (>400 mb). If I just want to run calculations on my own datasets, I need to download everything. I like @XericZephyr idea to keep the data in environment like OpenAI gym

**Comment by rbharath** [comment]
Here's a first iteration at a potential design

- Add a `dc.molnet` module to DeepChem. This module has a `run_benchmarks(datasets, model)` function. The `model` must be of type `dc.model.Model` (through inheritance). The `datasets` will be a list of strings naming the different datasets desired to be benchmarked against. Then users can run MoleculeNet benchmarks locally without having to contribute their code into the DeepChem repo. Users can use `xgboost`, or `pytorch` models locally without having to make a PR (or add dependencies) to DeepChem.
- To accept a new model as "official" (listed on the front benchmarks page), the model will need to be open source in a github/bitbucket repo. A DeepChem dev will code-review to make sure there's no cheating, then re-run the benchmark locally, and upload the numbers obtained.
- If users want to contribute a model to DeepChem, then follow the original procedure outlined above.

**Comment by XericZephyr** [comment]
LGTM!

I will be honored to be the first of the few guys to try out this new benchmark setting.

I will start working on it immediately and let you guys know my progress.

**Comment by rbharath** [comment]
I think the new `dc.molnet` module introduced in #437 has resolved this issue. Thanks for spurring the discussion @XericZephyr!
